File: client/archive/android/app/src/main/java/com/solutioncrafts/BriskVTU/core/models/Transaction.kt
package com.solutioncrafts.BriskVTU.core.models

import kotlinx.serialization.SerialName
import kotlinx.serialization.Serializable
import kotlinx.serialization.json.JsonElement

@Serializable
enum class TransactionType {
    @SerialName("TOP_UP") TOP_UP,
    @SerialName("PAYMENT") PAYMENT,
    @SerialName("CASHBACK") CASHBACK,
    @SerialName("REFUND") REFUND
}

@Serializable
enum class TransactionStatus {
    @SerialName("INITIATED") INITIATED,
    @SerialName("PENDING") PENDING,
    @SerialName("SUCCESS") SUCCESS,
    @SerialName("FAILED") FAILED
}

@Serializable
data class Transaction(
    val id: String,
    @SerialName("walletId") val walletId: String,
    @SerialName("userId") val userId: String,
    val type: TransactionType,
    val amount: Long, // in NGN Kobo
    val serviceFee: Long, // in NGN Kobo
    val currency: String = "NGN",
    val status: TransactionStatus,
    val exchangeRate: Double? = null,
    val failureReason: String? = null,
    val metadata: Map<String, JsonElement>? = null,
    @SerialName("createdAt") val createdAt: String,
    @SerialName("updatedAt") val updatedAt: String
) {
    val formattedAmount: String
        get() = "₦${String.format("%.2f", amount / 100.0)}"
        
    val formattedFee: String
        get() = "₦${String.format("%.2f", serviceFee / 100.0)}"
}

File: client/archive/ios/BriskVTU/Core/Models/Transaction.swift
import Foundation

enum TransactionType: String, Codable {
    case topUp = "TOP_UP"
    case payment = "PAYMENT"
    case cashback = "CASHBACK"
    case refund = "REFUND"
}

enum TransactionStatus: String, Codable {
    case initiated = "INITIATED"
    case pending = "PENDING"
    case success = "SUCCESS"
    case failed = "FAILED"
}

struct Transaction: Codable, Identifiable {
    let id: String
    let walletId: String
    let userId: String
    let type: TransactionType
    let amount: Int // in NGN Kobo
    let serviceFee: Int // in NGN Kobo
    let currency: String
    let status: TransactionStatus
    let exchangeRate: Double?
    let failureReason: String?
    let metadata: [String: AnyCodable]?
    let createdAt: Date
    let updatedAt: Date
    
    // Helper to format amount as currency
    var formattedAmount: String {
        let amountInNaira = Double(amount) / 100.0
        let formatter = NumberFormatter()
        formatter.numberStyle = .currency
        formatter.currencyCode = "NGN"
        formatter.currencySymbol = "₦"
        return formatter.string(from: NSNumber(value: amountInNaira)) ?? "₦\(String(format: "%.2f", amountInNaira))"
    }
}

// AnyCodable helper for heterogeneous dictionaries
struct AnyCodable: Codable {
    let value: Any

    init(_ value: Any) {
        self.value = value
    }

    init(from decoder: Decoder) throws {
        let container = try decoder.singleValueContainer()
        if let x = try? container.decode(Bool.self) {
            value = x
        } else if let x = try? container.decode(Int.self) {
            value = x
        } else if let x = try? container.decode(Double.self) {
            value = x
        } else if let x = try? container.decode(String.self) {
            value = x
        } else if let x = try? container.decode([String: AnyCodable].self) {
            value = x.mapValues { $0.value }
        } else if let x = try? container.decode([AnyCodable].self) {
            value = x.map { $0.value }
        } else {
            throw DecodingError.typeMismatch(AnyCodable.self, DecodingError.Context(codingPath: decoder.codingPath, debugDescription: "Wrong type for AnyCodable"))
        }
    }

    func encode(to encoder: Encoder) throws {
        var container = encoder.singleValueContainer()
        if let x = value as? Bool {
            try container.encode(x)
        } else if let x = value as? Int {
            try container.encode(x)
        } else if let x = value as? Double {
            try container.encode(x)
        } else if let x = value as? String {
            try container.encode(x)
        } else if let x = value as? [String: Any] {
            try container.encode(x.mapValues { AnyCodable($0) })
        } else if let x = value as? [Any] {
            try container.encode(x.map { AnyCodable($0) })
        } else {
            // Encode as empty if unknown
            try container.encodeNil()
        }
    }
}
